source/slang/glsl.meta.slang

```diff
@@ -2825,21 +2825,1989 @@ public uint rayQueryGetIntersectionTypeEXT(rayQueryEXT q, bool committed)
     return 0;
 }
 
+// TODO: implementation of built-in variables; proper tests; these are stubs
+// likley related to the following issue since GLSL adds new 
+// 'system' variables: https://github.com/shader-slang/slang/issues/411
+
+#define glsl_caps_shader_subgroup(param1) __glsl_extension(param1) [require(glsl)] 
+#define spirv_caps_shader_subgroup(param1) __spirv_version(param1) [require(spirv)] 
+
+//macros used to make implementation and modification reasonable under regex
+#define macroAnyVecGenericBuiltIn __generic<T : __BuiltinType, let N : int>
+#define macroAnyVecGenericLogical __generic<T : __BuiltinLogicalType, let N : int>
+#define macroAnyVecGenericArithmetic __generic<T : __BuiltinArithmeticType, let N : int>
+#define macroAnyVec vector<T, N>
+
+__generic<T : __BuiltinType>
+void shader_subgroup_preamble() {
+    // checks needed for shader_subgroup functions; __requireGLSLExtension does not work, the call fails to find the OP
+    __target_switch
+    {
+    case glsl:
+        if (__type_equals<T, half>()) __requireGLSLExtension("GL_EXT_shader_subgroup_extended_types_float16");
+        if (__type_equals<T, uint8_t>()) __requireGLSLExtension("GL_EXT_shader_subgroup_extended_types_int8");
+        if (__type_equals<T, uint16_t>()) __requireGLSLExtension("GL_EXT_shader_subgroup_extended_types_int16");
+        if (__type_equals<T, uint64_t>()) __requireGLSLExtension("GL_EXT_shader_subgroup_extended_types_int64");
+        return;
+    case spirv:
+    case cpp:
+    case cuda:
+    case cpp:
+        return;
+    }
+    return;
+} 
 
-//
-// Subgroup
-//
+// GL_KHR_shader_subgroup_basic Built-in Variables 
+
+glsl_caps_shader_subgroup(GL_KHR_shader_subgroup_basic)
+public uint gl_NumSubgroups;
```

Comment:
You said "in" or "out" modifiers are not a correct syntax for these.
But I see them on the other variables.
```
public out float4 gl_Position : SV_Position;
public out float gl_PointSize : SV_PointSize;
public in vec4 gl_FragCoord : SV_Position;
public out float gl_FragDepth : SV_Depth;
public out int gl_FragStencilRef : SV_StencilRef;
```
Are you sure you cannot use in/out modifiers on the subgroup variables?

Reply:
Is it hard to implement those variables properly?

Reply:
to implement the builtin's I would have to create new system variables since HLSL ones do not map over; in out requires the interface inheritance of a system variable 

The implementation of a system variable looks non trivial and quite unstructured in implementation. I noticed there was also an issue related to changing how defining them works at some level https://github.com/shader-slang/slang/issues/411